extensions/optaplanner/deployment/src/main/java/io/quarkus/optaplanner/deployment/OptaPlannerProcessor.java

```
@@ -0,0 +1,124 @@
+import org.optaplanner.core.api.score.stream.ConstraintProvider;
+
+import static io.quarkus.deployment.annotations.ExecutionTime.STATIC_INIT;
+
+class OptaPlannerProcessor {
+
+    @BuildStep
+    CapabilityBuildItem capability() {
+        return new CapabilityBuildItem(Capabilities.OPTAPLANNER);
+    }
+
+    @BuildStep
+    FeatureBuildItem feature() {
+        return new FeatureBuildItem(FeatureBuildItem.OPTAPLANNER);
+    }
+
+    @BuildStep
+    void registerAdditionalBeans(BuildProducer<AdditionalBeanBuildItem> additionalBeans) {
+        // The bean encapsulating the SolverFactory
+        additionalBeans.produce(new AdditionalBeanBuildItem(SolverFactoryProvider.class));
+    }
+
+    @BuildStep
+    @Record(STATIC_INIT)
+    void recordSolverFactory(OptaPlannerRecorder recorder,
+            CombinedIndexBuildItem combinedIndex,
+            BuildProducer<BeanContainerListenerBuildItem> beanContainerListener) {
+        IndexView indexView = combinedIndex.getIndex();
+        String solutionClassName = findSolutionClassName(indexView);
+        List<String> entityClassNameList = findEntityClassNameList(indexView);
+        String constraintProviderClassName = findConstraintProviderClassName(indexView);
+        beanContainerListener
+                .produce(new BeanContainerListenerBuildItem(
+                        recorder.initializeSolverFactory(solutionClassName, entityClassNameList, constraintProviderClassName)));
+    }
+
+    private String findSolutionClassName(IndexView indexView) {
+        Collection<AnnotationInstance> annotationInstances = indexView.getAnnotations(DotNames.PLANNING_SOLUTION);
+        if (annotationInstances.size() > 1) {
+            throw new IllegalStateException("Multiple classes (" + convertAnnotationInstancesToString(annotationInstances)
+                    + ") found with a @" + PlanningSolution.class.getSimpleName() + " annotation.");
+        }
+        if (annotationInstances.isEmpty()) {
```

I think the processor can just exit early if there are no instances instead of failing, right?

---

Good question, let me think:

- If there is a @PlanningSolution but no @PlanningEntity it must definitely fail fast.
- Visa versa definitely too.
- If there is no @PlanningSolution and no @PlanningEntity, but the optaplanner extension is in the dependency tree, should it fail fast or do nothing? Does quarkus has a policy on this? What does the rest module do if there are no REST methods?

---

Hibernate for example justs exists quickly if there are no entities. So it's really up to the extension to do what's best for its use case

---

I 'll follow hibernate's approach on such matters :)

---

Cool :)

---

Yes please.

---

@cuijulian ^^